modules/basis/basis-datasource/tomcat-listener/src/main/java/io/dirigible/tomcat/listener/DelegatingObjectFactory.java

```diff
@@ -0,0 +1,141 @@
+package io.dirigible.tomcat.listener;
+
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.naming.Context;
+import javax.naming.Name;
+import javax.naming.NameNotFoundException;
+import javax.naming.NamingException;
+import javax.naming.RefAddr;
+import javax.naming.Reference;
+import javax.naming.spi.ObjectFactory;
+
+public class DelegatingObjectFactory implements ObjectFactory {
+
+  private static final Logger LOGGER = Logger.getLogger(DelegatingObjectFactory.class.getName());
+
+  @Override
+  public Object getObjectInstance(Object obj, Name name, Context nameCtx, Hashtable<?, ?> environment) throws NamingException {
+    if (obj instanceof Reference) {
+      final Reference reference = (Reference) obj;
+      final String referenceType = reference.getClassName();
+      DelegatingObjectFactory.class.getClassLoader();
+      Object objectInstance = getObjectInstanceFromFactory(obj, name, nameCtx, environment, referenceType);
+      if (objectInstance != null) {
+        return objectInstance;
+      }
+
+      objectInstance = getObjectInstanceFromDefaults(obj, name, nameCtx, environment);
+      if (objectInstance != null) {
+        LOGGER.fine("Object instance is obtained from default factory.");
+        return objectInstance;
+      }
+    }
+    throw new NamingException("Object instance with name [" + name + "] could not be created.");
+  }
+
+  private Object getObjectInstanceFromFactory(Object obj, Name name, Context nameCtx, Hashtable<?, ?> environment,
+      final String referenceType) throws NamingException {
+    final List<ObjectFactory> factories = getObjectFactories(referenceType);
+    if (factories != null) {
+      for (ObjectFactory factory : factories) {
+        Object objectInstance = null;
+        try {
+          objectInstance = factory.getObjectInstance(obj, name, nameCtx, environment);
+        } catch (NameNotFoundException nne) {
+          LOGGER.fine("Will move to next factory as facroty [" + factory + "] could not retrieve object instance for name [" + name
+              + "]. NameNotFoundException was thrown. " + nne.getMessage());
+        } catch (Exception e) {
+          LOGGER.severe("Exception is thrown by facroty [" + factory + "] during retrieving object instance. Exception is: " + e);
+          NamingException ne = new NamingException("Cannot create resource  object instance due to exception in the object factory");
+          ne.initCause(e);
+          throw ne;
+        }
+        if (objectInstance != null) {
+          LOGGER.fine("Object instance with name [" + name + "]is created by factory [" + factory + "].");
+          return objectInstance;
+        }
+      }
```

Comment:
This logic is too complex to be nested. If the refactoring will take too long, create a separate issue for it.

Reply:
+1

Reply:
The logic is not complex it simply depends on a lot of catch blocks and loggers that add boilerplate, if you look closely the only method called is `objectInstance = factory.getObjectInstance(obj, name, nameCtx, environment);` everything else is defensive code